### ISSUE TITLE
SELinux context for init-script

### ISSUE THREAD

**Issue by cjeanneret** [EthACKdotOrg/orWall#79]
After many tests, it seems the init-script issue is due to SELinux context, NOT path as mentioned in #69.

In order to follow it, a new issue is created.

Here are the steps we tested on our own in order to debug this:
* Set init-script context to u:object_r:userinit_exec:s0 doesn't help (script doesn't even seem to be called)
* Set init-script context to u:r:init:s0 doesn't help (here again, script doesn't even seem to be called)
* "su" binary on CM12 comes from Superuser.apk, which doesn't support "--context" option
* Script doesn't find "su" binary at startup, even with an enforced PATH value
* We're able to run `setenforce 0` on CM12 with default "su" binary
* We might want to push a custom SELinux policy in order to allow the init-script to run iptables commands
* Apparently, iptables should be run in "u:r:init:s0" context, as tested by @ahf 
* this documentation might provide some hints: https://source.android.com/devices/tech/security/selinux/customize.html

What should the policy allow:
* run "iptables" binary in the "u:r:sysinit:s0" context
This is apparently the only authorization we might need — unsure about that.

**Comment by cjeanneret** [comment]
Something like this might be a start:

```SELinux
type orwall;
permissive_or_unconfined(orwall)
type orwall_exec, exec_type, file_type;

allow orwall shell_exec:file rx_file_perms;
allow orwall system_file:file rx_file_perms;
```

Though I have some doubts. Really not a SELinux admin :(.

**Comment by ahf** [comment]
Our troubles:

    type=2000 msg=audit(0.610:1): initialized
    type=1403 msg=audit(691415.530:2): policy loaded auid=4294967295 ses=4294967295
    type=1404 msg=audit(691415.550:3): enforcing=1 old_enforcing=0 auid=4294967295 ses=4294967295
    type=1400 msg=audit(691423.780:4): avc:  denied  { create } for  uid=0 pid=391 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691423.780:5): avc:  denied  { create } for  uid=0 pid=391 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691423.820:6): avc:  denied  { create } for  uid=0 pid=402 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691423.820:7): avc:  denied  { create } for  uid=0 pid=402 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691423.930:8): avc:  denied  { create } for  uid=0 pid=416 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691423.930:9): avc:  denied  { create } for  uid=0 pid=416 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691423.970:10): avc:  denied  { create } for  uid=0 pid=425 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691423.970:11): avc:  denied  { create } for  uid=0 pid=425 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691424.030:12): avc:  denied  { create } for  uid=0 pid=432 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691424.090:13): avc:  denied  { create } for  uid=0 pid=436 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691424.090:14): avc:  denied  { create } for  uid=0 pid=436 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691424.130:15): avc:  denied  { create } for  uid=0 pid=443 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691424.130:16): avc:  denied  { create } for  uid=0 pid=443 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
    type=1400 msg=audit(691424.170:17): avc:  denied  { create } for  uid=0 pid=451 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=1
    type=1400 msg=audit(691424.170:18): avc:  denied  { net_raw } for  uid=0 pid=451 comm="iptables" capability=13  scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=capability permissive=1
    type=1400 msg=audit(691424.170:19): avc:  denied  { getopt } for  uid=0 pid=451 comm="iptables" lport=255 scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=1
    type=1400 msg=audit(691424.170:20): avc:  denied  { net_admin } for  uid=0 pid=451 comm="iptables" capability=12  scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=capability permissive=1
    type=1400 msg=audit(691424.170:21): avc:  denied  { setopt } for  uid=0 pid=451 comm="iptables" lport=255 scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=1
    type=1400 msg=audit(691424.280:27): avc:  denied  { module_request } for  uid=0 pid=468 comm="ip6tables" kmod="ip6table_nat" scontext=u:r:sysinit:s0 tcontext=u:r:kernel:s0 tclass=system permissive=1

**Comment by ahf** [comment]
Odd that the permissive bit is changing during this.

**Comment by ahf** [comment]
I assume this is the policy that's loaded: https://android.googlesource.com/platform/external/sepolicy/+/master

**Comment by zerokes** [comment]
Hi, I was having the same problem I was manage to fix it by copying the file_contexts file to /data/security/current and modifying /system/bin/sysinit, /system/etc/init.d/90userinit, /data/local/userinit.sh lines to u:object_r:rootfs:s0. Then setenforce 0 to be able to set chcon for those files to that context and reload with restorecon the new selinux context file. Now init.d runs as root and can do whatever you want. It could also be changed to a more restrictive context to only allow iptables.

Regards,
Joaquín

**Comment by zerokes** [comment]
I'm testing and the userinit.sh context is dropped to init_shell since sysinit calls run-parts that transitions the context. One can change that context to rootfs or modify sysinit to call userinit.sh so it runs as init context.

**Comment by zerokes** [comment]
Deleted previous posts for a better solution. Using https://github.com/xmikos/setools-android/tree/master/jni/sepolicy-inject one can chang u:r:sysinit:s0 to permissive mode using command: 
sepolicy-inject -Z sysinit -P /sepolicy -o /data/tmp/sepolicy. Rules can be updated by copying selinux_version file, sepolicy, *_contexts and /system/etc/security/mac_permissions.xml to /data/security/current/ folder and setting setprop selinux.reload_policy 1.

**Comment by sniperle** [comment]
Yes, SELinux is the reason why /data/local/userinit.sh and /system/etc/init.d/90userinit aren't executed:

type=1400 msg=audit(797603.940:4): avc:  denied  { relabelfrom } for  pid=1 comm="init" name="userinit.sh" dev="mmcblk0p35" ino=580015 scontext=u:r:init:s0 tcontext=u:object_r:userinit_exec:s0 tclass=file permissive=0

type=1400 msg=audit(1441975252.903:910): avc:  denied  { getattr } for  pid=2918 comm="ostsq.commander" path="/system/etc/init.d/90userinit" dev="mmcblk0p30" ino=486 scontext=u:r:untrusted_app:s0 tcontext=u:object_r:userinit_exec:s0 tclass=file permissive=0
type=1400 msg=audit(1441975252.903:911): avc:  denied  { getattr } for  pid=2918 comm="ostsq.commander" path="/system/etc/init.d/90userinit" dev="mmcblk0p30" ino=486 scontext=u:r:untrusted_app:s0 tcontext=u:object_r:userinit_exec:s0 tclass=file permissive=0
type=1400 msg=audit(1441975252.903:912): avc:  denied  { getattr } for  pid=2918 comm="ostsq.commander" path="/system/etc/init.d/90userinit" dev="mmcblk0p30" ino=486 scontext=u:r:untrusted_app:s0 tcontext=u:object_r:userinit_exec:s0 tclass=file permissive=0

**Comment by sniperle** [comment]
I was able to set the correct SELinux rights to get userinit.sh executed

chcon u:object_r:userinit_exec:s0 /system/etc/init.d/90userinit
chcon u:object_r:userinit_data_exec:s0 /data/local/userinit.sh

**Comment by cjeanneret** [comment]
Heya!
thanks for the tip, I was thinking about that too, but due to some lack of time, I couldn't do it.
There's just a small problem: 90userinit and userinit.sh are actually system files (at least on CM12), hence their right will most probably be overridden with updates.

I'll try to find the time to set the context on our file instead of the userinit, though I'm pretty sure this won't work as expected… OrWall will need to check the script context and reset it — and then ask for a reboot. Humpf…

**Comment by sniperle** [comment]
Glad I could help. :-)

But nevertheless, if you try to execute iptables via userinit.sh, it's execution is denied.

**Comment by cjeanneret** [comment]
Ah, so that's the same error I currently have… In fact, on my CM12, init-script is launched, but I get the denied described in the previous messages. It seems I could use a "su" option in order to enforce command context, but this means relying on a precise "su" APK, which is kind of ugly, especially when there are multiple "su" APK, each ROM having its favorite :disappointed: 

Thus, it's a bit annoying. I'll try to find some time in order to sort this issue, as it's *really* a show stopper: we can't trust our device, hence we have to lock down its communications as soon as possible (which is the goal of the init-script: just lock down connections, until orWall is started).

There are two ways:
* either use the su capability when available (but as said, ugly as hell)
* or add a SELinux policy, which is the best way — though also a bit complicated…

The second way being the best, I'll need to find out what to push in it, and how to push the policy in the system. I read some weeks (or was it months??) ago some docs about it, but I'm really short of time…

**Comment by c3ph3us** [comment]
http://selinuxproject.org/page/NB_SEforAndroid_1#Updating_Policy_Files

http://seandroid.bitbucket.org/AddressingHiddenDenials.html#13

**Comment by infinity0** [comment]
Hey, what's the status on this? Running the following in an adb root shell, then rebooting, appears to work for me on CM-13:

~~~~
# cp /data/data/org.ethack.orwall/app_bin/userinit.sh /data/local
# chcon u:object_r:userinit_data_exec:s0 /data/local/userinit.sh
~~~~

Apps not enabled in orwall don't get any internet access, and this persists even if I disable then enable orwall. However, orwall still says "iptables appears not to be initialised".

In other words, we don't actually need our `su` to have a `--context` option, we can just run `chcon` as root. Since orwall requires su already, this shouldn't be a problem, right?

**Comment by infinity0** [comment]
Ah sorry for the noise - I'm running into the same issues described above, with iptables being denied execution. I tried to change `userinit.sh` to use `runcon` instead:

~~~~
IP6TABLES=/system/bin/runcon u:r:init:s0 /system/bin/ip6tables
IPTABLES=/system/bin/runcom u:r:init:s0 /system/bin/iptables
~~~~

but this just gets ignored and my logs show that it's still being run as `scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0`. This suggests that using `su --context` even if available, possibly wouldn't work anyway, since it typically just delegates to `runcon` - https://su.chainfire.eu/#selinux-contexts-switching-how - never mind the fact that superSU is closed source and could be sending all your data to whoever it wants.

I guess the behaviour I observed about apps getting blocked was just orwall working normally.

**Comment by isislovecruft** [comment]
Hey, I think I've gotten this working on device running Android 5.2.1.

The first thing I tried, **which did not work** was to do the following (on the device, or through an adb shell):
```
su
mount -o rw,remount /system
chcon u:object_r:userinit_exec:s0 /system/etc/init.d/90userinit
cp /data/data/org.ethack.orwall/app_bin/userinit.sh /data/local/userinit.sh
chcon u:object_t:userinit_data_exec:s0 /data/local/userinit.sh
```

This didn't seem to have any effect, but it seems wise to put it here, just in case the it did have some effect in combination with the things I tried on the second attempt. Doing the following (again, on the device) and then rebooting it seems to fix the issue:
```
su
mount -o rw,remount /system
mkdir /system/su.d
chmod 0700 /system/su.d
cp /etc/init.d/91firewall /system/su.d/
```

**Comment by ahf** [comment]
Cool!

Does `/system/su.d/` init scripts run before/together with/after the regular `/etc/init.d/` scripts?

**Comment by infinity0** [comment]
Which version of su do you have @isislovecruft? It sounds like `/system/su.d` is a SuperSU-specific [thing](https://su.chainfire.eu/#selinux-policies-supolicy) and appears to be ignored by other versions of `su`. But perhaps we can just hook into android`s `/init.*` [scripts](https://android.googlesource.com/platform/system/core/+/master/init/readme.txt) directly.

(As a side note, for security one should generally do `mkdir -m` instead of `mkdir; chmod`)

**Comment by infinity0** [comment]
lol, so one very dirty hack that Worked For Me on Nexus 5X / CM 13 is `chcon u:object_r:netd_exec /system/bin/sysinit`. I noticed from `adb logcat` that [netd](https://android.googlesource.com/platform/system/netd/) was running iptables fine during boot. A few caveats:

- this will disable `90userinit` since netd isn't allowed to run that, but if you need to, you can fix it with `chcon u:object_r:system_file:s0 /system/etc/init.d/90userinit /data/local/userinit.sh`. you can use `restorecon` to restore them later, which knows the original values.
- this effectively bypasses default android security policy which is not a good idea
- all of the above means it's not a suitable solution for orwall to adopt :(

Probably the best solution is to write a policy similar to [that for netd](https://android.googlesource.com/platform/external/sepolicy/+/master/netd.te). But even pretending that I've already done that, I still haven't managed to figure out how to actually *change* the current policy. It appears centralised in `/sepolicy` but my phone doesn't even have the tools to edit that type of file. The instructions linked to in above comments, are all about updating policy files *from a development machine* using source `.te` files which of course isn't suitable here.

Another potential option, if `ndc` supports the equivalent of what orwall is doing in `91firewall` and `sysinit` is allowed to call `ndc`, then we could rewrite it without worrying about selinux. (`ndc` communicates with `netd` which is allowed to run `iptables`.)

(We also need to check that `sysinit` exists on stock android; on my phone it's only referred to by `/init.cm.rc` which obviously is CM-only.)

**Comment by infinity0** [comment]
A much cleaner hack this time, after I [noticed this](https://android.googlesource.com/platform/system/netd/+/master/server/NetdConstants.cpp#34):

~~~~
$ su
# mount -o rw,remount /system
# cp /etc/init.d/91firewall /system/bin/oem-iptables-init.sh
~~~~

Reboot and orwall appears to have initialised the tables correctly. `91firewall` is still run but gets blocked by the default policy, as you can see in `logcat`. (Optionally, you can add the line `log -p i -t orwall "Executing orwall initscript"` to the beginning of the script. On my phone it runs ~2s after `91firewall`. Not sure if this is "early enough".)

I don't know if this would translate into a good solution for orwall... At least cyanogenmod (and I would guess stock android) is not already using `oem-iptables-init.sh`, but I don't know about other vendors. Looking at the git logs, this mechanism seems to have been around since Android 4. (Also, in `oem_iptables_hook.cpp` it looks like `netd` expects the rules to begin with `oem_` to be able to clean them up properly; I don't know how orwall would interact with that.)

But the advantage is you don't have to mess with selinux.

**Comment by patcon** [comment]
cc: @mikeperry-tor

**Comment by unclebob243** [comment]
This script works fine in LinageOS 14.1 latest.
```
#!/system/bin/sh
## Find your Orbot UID and put into variable ORBOT_UID
ORBOT_UID=10112
## Instal this script using root shell
##  mount -o remount,rw /system
##  cp /sdcard/oem-iptables-init.sh /system/bin/
##  chown root:root /system/bin/oem-iptables-init.sh
##  chmod 750 /system/bin/oem-iptables-init.sh
##  mount -o remount,ro /system
##  reboot
IP6TABLES=/system/bin/ip6tables
IPTABLES=/system/bin/iptables
IPTABLES_OPTS='--verbose'
## check if iptables can lock
$IPTABLES --help | grep -q -- --wait && IPTABLES_OPTS="$IPTABLES_OPTS --wait"
## check if initialized
$IPTABLES $IPTABLES_OPTS -C ow_OUTPUT_LOCK -j DROP 2>/dev/null && exit 0
## OUTPUT 
$IPTABLES $IPTABLES_OPTS -P OUTPUT DROP
$IPTABLES $IPTABLES_OPTS -N ow_OUTPUT_LOCK
$IPTABLES $IPTABLES_OPTS -A ow_OUTPUT_LOCK -m owner --uid-owner $ORBOT_UID -m conntrack --ctstate NEW,RELATED,ESTABLISHED -j ACCEPT
$IPTABLES $IPTABLES_OPTS -A ow_OUTPUT_LOCK -j DROP
$IPTABLES $IPTABLES_OPTS -I OUTPUT -j ow_OUTPUT_LOCK
## INPUT  
$IPTABLES $IPTABLES_OPTS -P INPUT DROP
$IPTABLES $IPTABLES_OPTS -N ow_INPUT_LOCK
$IPTABLES $IPTABLES_OPTS -A ow_INPUT_LOCK -m owner --uid-owner $ORBOT_UID -m conntrack --ctstate NEW,RELATED,ESTABLISHED -j ACCEPT
$IPTABLES $IPTABLES_OPTS -A ow_INPUT_LOCK -j DROP
$IPTABLES $IPTABLES_OPTS -I INPUT -j ow_INPUT_LOCK
## Block IPv6 traffic at boot
$IP6TABLES $IPTABLES_OPTS -P INPUT DROP
$IP6TABLES $IPTABLES_OPTS -P OUTPUT DROP
$IP6TABLES $IPTABLES_OPTS -P FORWARD DROP
$IP6TABLES $IPTABLES_OPTS -I INPUT -j REJECT
$IP6TABLES $IPTABLES_OPTS -I OUTPUT -j REJECT
$IP6TABLES $IPTABLES_OPTS -I FORWARD -j REJECT
##EOF
```
One issue with ORBOT_UID - you have to define it manually, due to SELinux does not allow to call 
```
ORBOT_UID=$(cat /data/system/packages.list | sed -n 's/^org.torproject.android //p' | cut -d ' ' -f1)
```
in this script.

Also this command:
```
settings put global http_proxy 127.0.0.1:8118 
```
wll help you make all apps using android http framefork to go through TOR, only permit them in OrWall to Localhost. Apps like Telegram, K9Mail using direct socket IO need to be transproxied.


Hope this helps.

**Comment by unclebob243** [comment]
Interesting case described here https://github.com/ferhung/Skip_network_validation
and it affects internet connectivity, unless used settings put global http_proxy 127.0.0.1:8118 and blocking google

**Comment by sawyer658** [comment]
jeanneret opened this issue on Mar 26, 2015 · 21 comments
Comments
@cjeanneret
 
Contributor
cjeanneret commented on Mar 26, 2015
After many tests, it seems the init-script issue is due to SELinux context, NOT path as mentioned in #69.

In order to follow it, a new issue is created.

Here are the steps we tested on our own in order to debug this:

Set init-script context to u:object_r:userinit_exec:s0 doesn't help (script doesn't even seem to be called)
Set init-script context to u:r:init:s0 doesn't help (here again, script doesn't even seem to be called)
"su" binary on CM12 comes from Superuser.apk, which doesn't support "--context" option
Script doesn't find "su" binary at startup, even with an enforced PATH value
We're able to run setenforce 0 on CM12 with default "su" binary
We might want to push a custom SELinux policy in order to allow the init-script to run iptables commands
Apparently, iptables should be run in "u:r:init:s0" context, as tested by @ahf
this documentation might provide some hints: https://source.android.com/devices/tech/security/selinux/customize.html
What should the policy allow:

run "iptables" binary in the "u:r:sysinit:s0" context
This is apparently the only authorization we might need — unsure about that.
@cjeanneret
 
Contributor
Author
cjeanneret commented on Mar 26, 2015
Something like this might be a start:

type orwall;
permissive_or_unconfined(orwall)
type orwall_exec, exec_type, file_type;

allow orwall shell_exec:file rx_file_perms;
allow orwall system_file:file rx_file_perms;
Though I have some doubts. Really not a SELinux admin :(.

@ahf
 
Contributor
ahf commented on Mar 26, 2015
Our troubles:

type=2000 msg=audit(0.610:1): initialized
type=1403 msg=audit(691415.530:2): policy loaded auid=4294967295 ses=4294967295
type=1404 msg=audit(691415.550:3): enforcing=1 old_enforcing=0 auid=4294967295 ses=4294967295
type=1400 msg=audit(691423.780:4): avc:  denied  { create } for  uid=0 pid=391 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691423.780:5): avc:  denied  { create } for  uid=0 pid=391 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691423.820:6): avc:  denied  { create } for  uid=0 pid=402 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691423.820:7): avc:  denied  { create } for  uid=0 pid=402 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691423.930:8): avc:  denied  { create } for  uid=0 pid=416 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691423.930:9): avc:  denied  { create } for  uid=0 pid=416 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691423.970:10): avc:  denied  { create } for  uid=0 pid=425 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691423.970:11): avc:  denied  { create } for  uid=0 pid=425 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691424.030:12): avc:  denied  { create } for  uid=0 pid=432 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691424.090:13): avc:  denied  { create } for  uid=0 pid=436 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691424.090:14): avc:  denied  { create } for  uid=0 pid=436 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691424.130:15): avc:  denied  { create } for  uid=0 pid=443 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691424.130:16): avc:  denied  { create } for  uid=0 pid=443 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=0
type=1400 msg=audit(691424.170:17): avc:  denied  { create } for  uid=0 pid=451 comm="iptables" scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=1
type=1400 msg=audit(691424.170:18): avc:  denied  { net_raw } for  uid=0 pid=451 comm="iptables" capability=13  scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=capability permissive=1
type=1400 msg=audit(691424.170:19): avc:  denied  { getopt } for  uid=0 pid=451 comm="iptables" lport=255 scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=1
type=1400 msg=audit(691424.170:20): avc:  denied  { net_admin } for  uid=0 pid=451 comm="iptables" capability=12  scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=capability permissive=1
type=1400 msg=audit(691424.170:21): avc:  denied  { setopt } for  uid=0 pid=451 comm="iptables" lport=255 scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0 tclass=rawip_socket permissive=1
type=1400 msg=audit(691424.280:27): avc:  denied  { module_request } for  uid=0 pid=468 comm="ip6tables" kmod="ip6table_nat" scontext=u:r:sysinit:s0 tcontext=u:r:kernel:s0 tclass=system permissive=1
@ahf
 
Contributor
ahf commented on Mar 26, 2015
Odd that the permissive bit is changing during this.

@ahf
 
Contributor
ahf commented on Mar 26, 2015
I assume this is the policy that's loaded: https://android.googlesource.com/platform/external/sepolicy/+/master

@cjeanneret cjeanneret referenced this issue on Mar 28, 2015
Init script issue on CM-12 #69
 Closed
@mikeperry-tor mikeperry-tor referenced this issue on Apr 9, 2015
Update script incompatible with Cyanogenmod 12's update-binary #50
 Open
@xeroxnir
 
xeroxnir commented on Jul 7, 2015
Deleted previous posts for a better solution. Using https://github.com/xmikos/setools-android/tree/master/jni/sepolicy-inject one can chang sysinit domain to permissive mode using command:
sepolicy-inject -Z sysinit -P /sepolicy -o /data/tmp/sepolicy. Rules can be updated by copying selinux_version file, sepolicy, *_contexts and /system/etc/security/mac_permissions.xml to /data/security/current/ folder and setting setprop selinux.reload_policy 1.

@cjeanneret cjeanneret referenced this issue on Aug 12, 2015
"Iptables doesn't seem to be initialized" #84
 Closed
@DocSniper
 
DocSniper commented on Sep 11, 2015
Yes, SELinux is the reason why /data/local/userinit.sh and /system/etc/init.d/90userinit aren't executed:

type=1400 msg=audit(797603.940:4): avc: denied { relabelfrom } for pid=1 comm="init" name="userinit.sh" dev="mmcblk0p35" ino=580015 scontext=u:r:init:s0 tcontext=u:object_r:userinit_exec:s0 tclass=file permissive=0

type=1400 msg=audit(1441975252.903:910): avc: denied { getattr } for pid=2918 comm="ostsq.commander" path="/system/etc/init.d/90userinit" dev="mmcblk0p30" ino=486 scontext=u:r:untrusted_app:s0 tcontext=u:object_r:userinit_exec:s0 tclass=file permissive=0
type=1400 msg=audit(1441975252.903:911): avc: denied { getattr } for pid=2918 comm="ostsq.commander" path="/system/etc/init.d/90userinit" dev="mmcblk0p30" ino=486 scontext=u:r:untrusted_app:s0 tcontext=u:object_r:userinit_exec:s0 tclass=file permissive=0
type=1400 msg=audit(1441975252.903:912): avc: denied { getattr } for pid=2918 comm="ostsq.commander" path="/system/etc/init.d/90userinit" dev="mmcblk0p30" ino=486 scontext=u:r:untrusted_app:s0 tcontext=u:object_r:userinit_exec:s0 tclass=file permissive=0

@DocSniper
 
DocSniper commented on Sep 12, 2015
I was able to set the correct SELinux rights to get userinit.sh executed

chcon u:object_r:userinit_exec:s0 /system/etc/init.d/90userinit
chcon u:object_r:userinit_data_exec:s0 /data/local/userinit.sh

@cjeanneret
 
Contributor
Author
cjeanneret commented on Sep 14, 2015
Heya!
thanks for the tip, I was thinking about that too, but due to some lack of time, I couldn't do it.
There's just a small problem: 90userinit and userinit.sh are actually system files (at least on CM12), hence their right will most probably be overridden with updates.

I'll try to find the time to set the context on our file instead of the userinit, though I'm pretty sure this won't work as expected… OrWall will need to check the script context and reset it — and then ask for a reboot. Humpf…

@DocSniper
 
DocSniper commented on Sep 14, 2015
Glad I could help. :-)

But nevertheless, if you try to execute iptables via userinit.sh, it's execution is denied.

@cjeanneret
 
Contributor
Author
cjeanneret commented on Sep 15, 2015
Ah, so that's the same error I currently have… In fact, on my CM12, init-script is launched, but I get the denied described in the previous messages. It seems I could use a "su" option in order to enforce command context, but this means relying on a precise "su" APK, which is kind of ugly, especially when there are multiple "su" APK, each ROM having its favorite disappointed

Thus, it's a bit annoying. I'll try to find some time in order to sort this issue, as it's really a show stopper: we can't trust our device, hence we have to lock down its communications as soon as possible (which is the goal of the init-script: just lock down connections, until orWall is started).

There are two ways:

either use the su capability when available (but as said, ugly as hell)
or add a SELinux policy, which is the best way — though also a bit complicated…
The second way being the best, I'll need to find out what to push in it, and how to push the policy in the system. I read some weeks (or was it months??) ago some docs about it, but I'm really short of time…

@c3ph3us
 
c3ph3us commented on Sep 16, 2015
http://selinuxproject.org/page/NB_SEforAndroid_1#Updating_Policy_Files

http://seandroid.bitbucket.org/AddressingHiddenDenials.html#13

@duud duud referenced this issue on Oct 1, 2015
connection not blocked #440
 Closed
@c3ph3us c3ph3us referenced this issue on Oct 7, 2015
add se policy for orwall to init iptables #1
 Open
@infinity0
 
infinity0 commented on Jan 2, 2016
Hey, what's the status on this? Running the following in an adb root shell, then rebooting, appears to work for me on CM-13 (whose su does not have --context):

# cp /data/data/org.ethack.orwall/app_bin/userinit.sh /data/local
# chcon u:object_r:userinit_data_exec:s0 /data/local/userinit.sh
Apps not enabled in orwall don't get any internet access, and this persists even if I disable then enable orwall. However, orwall still says "iptables appears not to be initialised".

In other words, we don't actually need our su to have a --context option, we can just run chcon as root. Since orwall requires su already, this shouldn't be a problem, right?

@infinity0
 
infinity0 commented on Jan 2, 2016
Ah sorry for the noise - I'm running into the same issues described above, with iptables being denied execution. I tried to change userinit.sh to use runcon instead:

IP6TABLES=/system/bin/runcon u:r:init:s0 /system/bin/ip6tables
IPTABLES=/system/bin/runcom u:r:init:s0 /system/bin/iptables
but this just gets ignored and my logs show that it's still being run as scontext=u:r:sysinit:s0 tcontext=u:r:sysinit:s0. This suggests that using su --context even if available, possibly wouldn't work anyway, since it typically just delegates to runcon - https://su.chainfire.eu/#selinux-contexts-switching-how - never mind the fact that superSU is closed source and could be sending all your data to whoever it wants.

I guess the behaviour I observed about apps getting blocked was just orwall working normally.

@isislovecruft
 
isislovecruft commented on Jan 25, 2016
Hey, I think I've gotten this working on device running Android 5.2.1.

The first thing I tried, which did not work was to do the following (on the device, or through an adb shell):

su
mount -o rw,remount /system
chcon u:object_r:userinit_exec:s0 /system/etc/init.d/90userinit
cp /data/data/org.ethack.orwall/app_bin/userinit.sh /data/local/userinit.sh
chcon u:object_t:userinit_data_exec:s0 /data/local/userinit.sh
This didn't seem to have any effect, but it seems wise to put it here, just in case the it did have some effect in combination with the things I tried on the second attempt. Doing the following (again, on the device) and then rebooting it seems to fix the issue:

su
mount -o rw,remount /system
mkdir /system/su.d
chmod 0700 /system/su.d
cp /etc/init.d/91firewall /system/su.d/
@ahf
 
Contributor
ahf commented on Jan 25, 2016
Cool!

Does /system/su.d/ init scripts run before/together with/after the regular /etc/init.d/ scripts?

@infinity0
 
infinity0 commented on Jan 25, 2016
Which version of su do you have @isislovecruft? It sounds like /system/su.d is a SuperSU-specific thing and appears to be ignored by other versions of su. But perhaps we can just hook into android's /init.* scripts directly.

(As a side note, for security one should generally do mkdir -m instead of mkdir; chmod)

@infinity0
 
infinity0 commented on Jan 25, 2016
lol, so one very dirty hack that Worked For Me on Nexus 5X / CM 13 is chcon u:object_r:netd_exec /system/bin/sysinit. I noticed from adb logcat that netd was running iptables fine during boot. A few caveats:

this will disable 90userinit since netd isn't allowed to run that, but if you need to, you can fix it with chcon u:object_r:system_file:s0 /system/etc/init.d/90userinit /data/local/userinit.sh. you can use restorecon to restore them later, which knows the original values.
this effectively bypasses default android security policy which is not a good idea
all of the above means it's not a suitable solution for orwall to adopt :(
Probably the best solution is to write a policy similar to that for netd. But even pretending that I've already done that, I still haven't managed to figure out how to actually change the current policy. It appears centralised in /sepolicy but my phone doesn't even have the tools to edit that type of file. The instructions linked to in above comments, are all about updating policy files from a development machine using source .te files which of course isn't suitable here. (edit: oh but now I see that's why @xeroxnir linked sepolicy-inject )

Another potential option, if ndc supports the equivalent of what orwall is doing in 91firewall and sysinit is allowed to call ndc, then we could rewrite it without worrying about selinux. (ndc communicates with netd which is allowed to run iptables.)

(We also need to check that sysinit exists on stock android; on my phone it's only referred to by /init.cm.rc which obviously is CM-only.)

@infinity0
 
infinity0 commented on Jan 25, 2016
A much cleaner hack this time, after I noticed this:

$ su
# mount -o rw,remount /system
# cp /etc/init.d/91firewall /system/bin/oem-iptables-init.sh
Then use nano to add the line log -p i -t orwall "Executing orwall initscript" to the beginning of the copied script. Reboot and watch logcat - 91firewall is still run but gets blocked by the default policy, then you should see the "orwall" log message (on my phone it runs ~2s after 91firewall) and a short while later "OemIptablesHook: OEM iptable hook installed". If you go into the orWall app, there hopefully should be no warning message about iptables not initialized.

I don't know if this would translate into a good solution for orwall... At least cyanogenmod (and I would guess stock android) is not already using oem-iptables-init.sh, but I don't know about other vendors. Looking at the git logs, this mechanism seems to have been around since Android 4. (Also, in oem_iptables_hook.cpp it looks like netd expects the rules to begin with oem_ to be able to clean them up properly; I don't know how orwall would interact with that.)

But the advantage is you don't have to mess with selinux.

@patcon
 
patcon commented on Jan 25, 2016
cc: @mikeperry-tor

EDIT: I'm creeping here, but this seems awesome. Y'all are great

@infinity0 infinity0 referenced this issue on Jan 25, 2016
orwall prevents internet detection when using wifi #87
 Open
@mikeperry-tor mikeperry-tor referenced this issue on Oct 27, 2016
1.2.0 broke oem-iptables-init.sh #121
 Open
@unclebob243
 
unclebob243 commented on May 26, 2017
This script works fine in LinageOS 14.1 latest.

#!/system/bin/sh
## Find your Orbot UID and put into variable ORBOT_UID
ORBOT_UID=10112
## Instal this script using root shell
##  mount -o remount,rw /system
##  cp /sdcard/oem-iptables-init.sh /system/bin/
##  chown root:root /system/bin/oem-iptables-init.sh
##  chmod 750 /system/bin/oem-iptables-init.sh
##  mount -o remount,ro /system
##  reboot
